src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DebugDirectoryEntryNode.cs

```diff
@@ -0,0 +1,164 @@
+            sb.Append($"__NativeRvaBlob_{_module.Assembly.GetName().Name}");
+        }
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            ObjectDataBuilder builder = new ObjectDataBuilder(factory, relocsOnly);
+            builder.RequireInitialPointerAlignment();
+            builder.AddSymbol(this);
+
+            // Emit empty entry. This will be filled with data after the output image is emitted
+            builder.EmitZeros(RSDSSize);
+
+            return builder.ToObjectData();
+        }
+
+        public byte[] GenerateRSDSEntryData(byte[] md5Hash)
+        {
+            MemoryStream rsdsEntry = new MemoryStream(RSDSSize);
+
+            using (BinaryWriter writer = new BinaryWriter(rsdsEntry))
+            {
+                // Magic "RSDS"
+                writer.Write((uint)0x53445352);
+
+                // The PDB signature will be the same as our NGEN signature.
+                // However we want the printed version of the GUID to be the same as the
+                // byte dump of the signature so we swap bytes to make this work.
+                Debug.Assert(md5Hash.Length == 16);
+                writer.Write((uint)((md5Hash[0] * 256 + md5Hash[1]) * 256 + md5Hash[2]) * 256 + md5Hash[3]);
+                writer.Write((ushort)(md5Hash[4] * 256 + md5Hash[5]));
+                writer.Write((ushort)(md5Hash[6] * 256 + md5Hash[7]));
+                writer.Write(md5Hash, 8, 8);
+
+                // Age
+                writer.Write(1);
+
+                string pdbFileName = _module.Assembly.GetName().Name + ".ni.pdb";
+                byte[] pdbFileNameBytes = Encoding.UTF8.GetBytes(pdbFileName);
+                writer.Write(pdbFileNameBytes);
+
+                Debug.Assert(rsdsEntry.Length <= RSDSSize);
+                return rsdsEntry.ToArray();
+            }
+        }
+    }
+
+    public class CopiedDebugDirectoryEntryNode : DebugDirectoryEntryNode
+    {
+        private readonly int _debugEntryIndex;
+
+        public override int ClassCode => 1558397;
+
+        public CopiedDebugDirectoryEntryNode(EcmaModule sourceModule, int debugEntryIndex)
+            : base(sourceModule)
+        {
+            Debug.Assert(debugEntryIndex >= 0);
+            _debugEntryIndex = debugEntryIndex;
+        }
+
+        public override void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix);
+            sb.Append($"__CopiedDebugEntryNode_{_debugEntryIndex}_{_module.Assembly.GetName().Name}");
+        }
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            if (relocsOnly)
+            {
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolDefinitionNode[] { this });
+            }
+
+            ObjectDataBuilder builder = new ObjectDataBuilder(factory, relocsOnly);
```

Comment:
Nit: `ObjectDataBuilder` is unnecessary if we're not building anything. This could just be `return new ObjectData(result, Array.Empty<Relocation>(), _module.Context.Target.PointerSize, new ISymbolDefinitionNode[] { this });` at the end (and similar for the node above).